### ISSUE TITLE
Add rule to warn if frozen dataclass gets mutable type annotation

### ISSUE THREAD

**Issue by Sinthoras7** [astral-sh/ruff#28221]
### Summary

I want to have the opt-in-option to get a warning if even the type annotation of an attribute allows for mutating them.

For example, currently, if I write
```python
@dataclass(frozen=True)
class CalculationResult:
   resulting_numbers: list[float]
```
nothing warns me that the immutability of an object `CalculationResult` is quite shallow and it can be modified. 

Ruff already has the `is_immutable_annotation` (used in e.g. [RUF008](https://docs.astral.sh/ruff/rules/mutable-dataclass-default/), so maybe that can be reused.

Another thing, why the current state is not ideal: With eq=True (the default) and frozen=True, dataclass generates __hash__ as hash((self.f1, ...)). Therefore, the class is also accepted by type checkers where Hashable is required. But when an attribute is an unhashable list, this gives a runtime error.
